File: SURVEYllance/Resources/Survey.cs
﻿using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;

namespace SURVEYllance.Resources
{
    //TODO: Add documentation
    public class Survey
    {
        #region Private properties
        
        /// <summary>
        /// Id to identify the survey from frontend calls
        /// </summary>
        private readonly string _id;
        
        /// <summary>
        /// The title of this survey
        /// </summary>
        private string _title;
        
        /// <summary>
        /// Possible answers
        /// </summary>
        private IReadOnlyList<SurveyAnswer> _answers;

        /// <summary>
        /// A survey can be closed, so the results will be displayed to anyone and it is not possible to vote anymore
        /// </summary>
        private bool _isClosed = false;

        #endregion

        #region Getter and Setter
        
        /// <summary>
        /// Getter for <see cref="_id"/>
        /// Id is generated on construction
        /// </summary>
        public string Id => _id;

        /// <summary>
        /// Getter of <see cref="_title"/>
        /// Title can only be set on construction
        /// </summary>
        public string Title
        {
            get => _title;
        }
        
        /// <summary>
        /// Getter of <see cref="_answers"/>
        /// Answers can only be set on construction
        /// </summary>
        public IReadOnlyList<SurveyAnswer> Answers
        {
            get => _answers;
        }
        
        /// <summary>
        /// Getter and Setter of <see cref="_isClosed"/>
        /// Questions can only be closed
        /// Fire event if is closed
        /// </summary>
        public bool IsClosed
        {
            get => _isClosed;
            set
            {
                if(!value) return;
                _isClosed = value;
                OnCloseSurvey?.Invoke();
            }
        }

        #endregion

        #region Constructor

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="title">Title of the survey. Usually a question</param>
        /// <param name="answers">Possible answers for this survey</param>
        public Survey(string title, IReadOnlyList<SurveyAnswer> answers)
        {
            _id = Convert.ToBase64String(Guid.NewGuid().ToByteArray());
            _title = title;
            _answers = answers;
            
            // Fire event if one vote for one answer has changed
            foreach (var surveyAnswer in answers)
            {
                surveyAnswer.OnVotesChange += () =>
                {
                    OnVotesChange?.Invoke(this, surveyAnswer);
                };
            }
        }

        #endregion

        #region methodes
        //TODO: Check if person already voted

        /// <summary>
        /// Vote for a specific answer
        /// Do not count if Survey is closed
        /// </summary>
        /// <param name="id">ID of the answer</param>
        public void VoteForAnswer(string id)
        {
            if(_isClosed) return;
            SurveyAnswer answer = _answers.FirstOrDefault(answer => answer.Id == id);
            if (answer == null) return;
            answer.Votes++;
        }
        
        /// <summary>
        /// Vote for an specific answer
        /// Do not count if Survey is closed
        /// </summary>
        /// <param name="answer">The answer-object</param>
        [Obsolete("Use VoteForAnswer(string id) instead")]
        public void VoteForAnswer(SurveyAnswer answer)
        {
            if(_isClosed) return;
            SurveyAnswer tempAnswer = _answers.FirstOrDefault(pAnswer => pAnswer == answer);
            if (tempAnswer != null)
            {
                tempAnswer.Votes++;
            }
        }

        #endregion
        
        #region Event handling

        // Delegate type for the event handler
        /// <summary>
        /// Delegate of <see cref="SurveyAnswer.OnVotesChange"/>
        /// </summary>
        /// <param name="answer">Tha answer that has been changed</param>
        public delegate void VotesChanged(Survey survey, SurveyAnswer answer);

        // Declare the event.
        /// <summary>
        /// Fire when the number of votes for this answer increases
        /// </summary>
        public event VotesChanged OnVotesChange;
        
        /// <summary>
        /// Delegate of <see cref="Survey.OnCloseSurvey"/>
        /// </summary>
        public delegate void CloseSurvey();

        /// <summary>
        /// Fire when the survey is closed
        /// </summary>
        public event CloseSurvey OnCloseSurvey;


        #endregion
    }
}

File: SURVEYllance/ClientApp/webroot/src/sharedObjects/Survey.js
/**
 * Survey-Object represents survey and it's answers.
 *  @property {string} Id id to identify the survey from backend calls
 *  @property {string} Title title of the survey
 *  @property {Array<SurveyAnswer>} Answers Array of {@link SurveyAnswer}
 *  @property {boolean} IsClosed Is the survey still running or closed?
 */
class Survey {

    //<editor-fold desc="Private properties">

    /**
     * id to identify the survey from backend calls
     * @type {string}
     * @private
     */
    #id;

    /**
     * title of this survey
     * @type {string}
     */
    #title;

    /**
     * Possible answers
     * @type {Array<SurveyAnswer>}
     */
    #answers;

    /**
     * A survey can be closed, so the results will be displayed to anyone and it is not possible to vote anymore
     * @type {boolean}
     */
    #isClosed;

    //</editor-fold>

    //<editor-fold desc="Getter and Setter">

    /**
     * Getter of {@link #id}
     * @returns {string}
     */
    get id() {
        return this.#id;
    }

    /**
     * Getter of {@link #title}
     * @returns {string}
     */
    get title() {
        return this.#title;
    }

    /**
     * Getter of {@link #answers}
     * @returns {Array<SurveyAnswer>}
     */
    get answers() {
        return this.#answers;
    }

    /**
     * Setter of {@link #id}
     * @param {string} value ID of the survey
     */
    set id(value) {
        this.#id = value;
    }

    /**
     * Setter of {@link #answers}
     * @param {Array<SurveyAnswer>} value New list of survey answers with updated values
     */
    set answers(value) {
        this.#answers = value;
    }

    /**
     * Getter of {@link #isClosed}
     * @returns {boolean}
     * @readonly
     */
    get isClosed() {
        return this.#isClosed;
    }

    /**
     * Setter of {@link #isClosed}
     * @param {boolean} value set survey closed
     */
    set isClosed(value) {
        if(!value)
            return;
        this.#isClosed = value;
        //TODO: Fire event
    }

    //</editor-fold>

    //<editor-fold desc="Constructor">

    /**
     * Constructor to create a survey
     * @param {string} title title of the survey, typically a question
     * @param {Array<SurveyAnswer>} surveyAnswers Array of answers. Must be of {@see surveyAnswers} Object
     * @constructor
     */
    constructor(title, surveyAnswers) {
        this.#title = title;
        this.#answers = surveyAnswers;
    }

    //</editor-fold>

    //<editor-fold desc="Public methods">

    /**
     * Vote for an answer
     * Do not count if Survey is closed
     * @param {string} answerId ID of the answer to vote for
     */
    VoteForAnswer(answerId){
        if(this.#isClosed)
            return;
        //Vote for answer
        let answer = this.#answers.find(answer => answer.id === answerId);
        answer.votes++;
    }

    /**
     * This method overrides the default toJSON method, and returns object WITH the private properties, so they will be passed to the Backend
     * @return {{isClosed: boolean, answers: Array<SurveyAnswer>, id: string, title: string}} Object with the private properties
     * @override
     */
    toJSON() {
        return {
            id: this.#id,
            title: this.#title,
            answers: this.#answers,
            isClosed: this.#isClosed
        };
    }

    //</editor-fold>

}